### ISSUE TITLE
Bad exit scenario

### ISSUE THREAD

**Issue by huntc** [landlord/landlord#76]
Observed the following when shutting down landlord:

```
Slf4jLogger started
Unix ready on /var/run/landlord/landlord.sock
An internal error has occurred within landlord given process 0. Stacktrace follows.
java.lang.InterruptedException
	at java.util.concurrent.locks.AbstractQueuedSynchronizer.doAcquireSharedNanos(AbstractQueuedSynchronizer.java:1039)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer.tryAcquireSharedNanos(AbstractQueuedSynchronizer.java:1328)
	at scala.concurrent.impl.Promise$DefaultPromise.tryAwait(Promise.scala:244)
	at scala.concurrent.impl.Promise$DefaultPromise.ready(Promise.scala:254)
	at scala.concurrent.impl.Promise$DefaultPromise.ready(Promise.scala:183)
	at scala.concurrent.Await$.$anonfun$ready$1(package.scala:191)
	at scala.concurrent.BlockContext$DefaultBlockContext$.blockOn(BlockContext.scala:53)
	at scala.concurrent.Await$.ready(package.scala:142)
	at akka.actor.CoordinatedShutdown$.$anonfun$initJvmHook$1(CoordinatedShutdown.scala:219)
	at akka.actor.CoordinatedShutdown$$anon$2.run(CoordinatedShutdown.scala:554)
Unable to access shutdown hooks; they will not be run until the JVM exits
```

Also noticed the following from one of the clients:

```
[INFO] [07/18/2018 08:47:00.203] [mqtt-gateway-akka.actor.default-dispatcher-2] [akka.actor.ActorSystemImpl(mqtt-gateway)] Server listening on /0.0.0.0:8083
[WARN] [07/18/2018 08:56:30.623] [mqtt-gateway-akka.actor.default-dispatcher-17] [RestartWithBackoffFlow(akka://mqtt-gateway)] Restarting graph due to failure. stack_trace: 
Connection lost (32109) - java.io.EOFException
	at org.eclipse.paho.client.mqttv3.internal.CommsReceiver.run(CommsReceiver.java:181)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.io.EOFException
	at java.io.DataInputStream.readByte(DataInputStream.java:267)
	at org.eclipse.paho.client.mqttv3.internal.wire.MqttInputStream.readMqttWireMessage(MqttInputStream.java:92)
	at org.eclipse.paho.client.mqttv3.internal.CommsReceiver.run(CommsReceiver.java:133)
	... 7 more

[ERROR] [07/18/2018 08:56:32.517] [mqtt-gateway-akka.actor.default-dispatcher-3] [akka.actor.ActorSystemImpl(mqtt-gateway)] mqttEventsProcess failed, exiting
java.lang.IllegalArgumentException: Unknown tailer: 0
	at com.github.huntc.streambed.durablequeue.chroniclequeue.ReaderWriter$$anonfun$handle$1.applyOrElse(ChronicleQueue.scala:408)
	at akka.actor.Actor.aroundReceive(Actor.scala:517)
	at akka.actor.Actor.aroundReceive$(Actor.scala:515)
	at com.github.huntc.streambed.durablequeue.chroniclequeue.ReaderWriter.akka$actor$Timers$$super$aroundReceive(ChronicleQueue.scala:281)
	at akka.actor.Timers.aroundReceive(Timers.scala:51)
	at akka.actor.Timers.aroundReceive$(Timers.scala:40)
	at com.github.huntc.streambed.durablequeue.chroniclequeue.ReaderWriter.aroundReceive(ChronicleQueue.scala:281)
	at akka.actor.ActorCell.receiveMessage(ActorCell.scala:588)
	at akka.actor.ActorCell.invoke(ActorCell.scala:557)
	at akka.dispatch.Mailbox.processMailbox(Mailbox.scala:258)
	at akka.dispatch.Mailbox.run(Mailbox.scala:225)
	at akka.dispatch.Mailbox.exec(Mailbox.scala:235)
	at akka.dispatch.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
	at akka.dispatch.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
	at akka.dispatch.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
	at akka.dispatch.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)

[ERROR] [07/18/2018 08:56:32.581] [mqtt-gateway-akka.actor.default-dispatcher-3] [akka.dispatch.Dispatcher] null
com.github.huntc.landlord.JvmExecutor$ExitException
	at com.github.huntc.landlord.JvmExecutor$$anonfun$starting$1$$anon$3.checkExit(JvmExecutor.scala:483)
	at com.github.huntc.landlord.ThreadGroupSecurityManager.checkExit(ThreadGroupMappings.scala:222)
	at java.lang.Runtime.exit(Runtime.java:107)
	at java.lang.System.exit(System.java:971)
	at com.github.huntc.streambed.mqttgateway.MqttGateway$.$anonfun$main$2(MqttGateway.scala:394)
	at com.github.huntc.streambed.mqttgateway.MqttGateway$.$anonfun$main$2$adapted(MqttGateway.scala:388)
	at scala.concurrent.impl.CallbackRunnable.run(Promise.scala:60)
	at akka.dispatch.BatchingExecutor$AbstractBatch.processBatch(BatchingExecutor.scala:55)
	at akka.dispatch.BatchingExecutor$BlockableBatch.$anonfun$run$1(BatchingExecutor.scala:91)
	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:12)
	at scala.concurrent.BlockContext$.withBlockContext(BlockContext.scala:81)
	at akka.dispatch.BatchingExecutor$BlockableBatch.run(BatchingExecutor.scala:91)
	at akka.dispatch.TaskInvocation.run(AbstractDispatcher.scala:40)
	at akka.dispatch.ForkJoinExecutorConfigurator$AkkaForkJoinTask.exec(ForkJoinExecutorConfigurator.scala:44)
	at akka.dispatch.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
	at akka.dispatch.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
	at akka.dispatch.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
	at akka.dispatch.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)
```

**Comment by longshorej** [comment]
The first exception is thrown by an application running in landlordd, see https://github.com/akka/akka/blob/v2.5.13/akka-actor/src/main/scala/akka/actor/CoordinatedShutdown.scala#L219 -- `InterruptedException` is not caught by `NonFatal` so it bubbles up.

Perhaps we should capture `InterruptedException` as they are to be expected when we interrupt an application's threads.

> Unable to access shutdown hooks; they will not be run until the JVM exits

This is spurious logging in this case -- the JVM is shutting down so the applications running in landlordd cannot access the shutdown hooks. Perhaps we should not log if we can determine the JVM is shutting down.

**Comment by dwijnand** [comment]
Apparently `InterruptedException`s "should be converted to an ExecutionException and propagated to the caller"

https://github.com/scala/bug/issues/8938

**Comment by longshorej** [comment]
Ah thanks for that link @dwijnand, some good info there. In our case, the caller is in the process of being unloaded / has already been so I don't think there's much we can do :stuck_out_tongue: 

I think this issue can be closed now that #81 is in, but going to let it sit for a bit for now.
